### PR TITLE
* Move precompiled assemblies into a resource in the primary assembly.

### LINKED ISSUE #1693: TagHelper and RazorPreCompileModule don't work in the same project and reference projects
I have a TagHelper in the web project itself (not as a reference) and I want to use RazorPreCompileModule in that project. This throws during compilation because the preprocess step tries to load the assembly that is currently created and cannot find it (yet).

### PR DESCRIPTION
**Pull request by pranavkm**
* Emit the primary assembly if a tag helper requires it.
* Make TagHelperSample.Web use precompilation.

Fixes #1693

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @yishaigalatzer

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/DefaultPrecompiledViewsProvider.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                       !typeInfo.ContainsGenericParameters;
+            }
+
+            return false;
+        }
+
+        private Assembly LoadPrecompiledViewsAssembly(RazorFileInfoCollection viewCollection)
+        {
+            var viewCollectionAssembly = viewCollection.GetType().GetTypeInfo().Assembly;
+            using (var assemblyStream = viewCollectionAssembly.GetManifestResourceStream(
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/DefaultPrecompiledViewsProvider.cs]: Should we throw here if the resource is not found in the assembly?

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs [hunk 2]
```diff
@@ -23,38 +21,26 @@ public class CompilerCache : ICompilerCache
         /// Initializes a new instance of <see cref="CompilerCache"/> populated with precompiled views
         /// discovered using <paramref name="provider"/>.
         /// </summary>
-        /// <param name="provider">
-        /// An <see cref="IAssemblyProvider"/> representing the assemblies
-        /// used to search for pre-compiled views.
-        /// </param>
+        /// <param name="viewsProvider"><see cref="IPrecompiledViewsProvider"/> that provides precompiled
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: this seems unrelated to this change

File: src/Microsoft.AspNet.Mvc.Razor/Internal/SymbolsUtility.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            if (PlatformHelper.IsMono)
+            {
+                return false;
+            }
+
+            try
+            {
+                // Check for the pdb writer component that roslyn uses to generate pdbs
+
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Razor/Internal/SymbolsUtility.cs]: remove empty line

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/DefaultPrecompiledViewsProvider.cs (hunk 1 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/DefaultPrecompiledViewsProvider.cs]: yes

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/DefaultPrecompiledViewsProvider.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                viewCollection.AssemblyResourceName))
+            {
+                Stream symbolsStream = null;
+                if (!string.IsNullOrEmpty(viewCollection.SymbolsResourceName))
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/DefaultPrecompiledViewsProvider.cs]: should we generate pdbs for release builds? Probably not I assume, because it will make the project assembly much larger

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/PrecompiledViewInfo.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNet.Mvc.Razor
+{
+    /// <summary>
+    /// Represents the result of <see cref="IPrecompiledViewsProvider"/>.
+    /// </summary>
+    public class PrecompiledViewInfo
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/PrecompiledViewInfo.cs]: i'm not quite following will this set of changes is necessary

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorFileInfoCollection.cs [hunk 6]
```diff
@@ -5,8 +5,24 @@
...
+
+        /// <summary>
+        /// Gets or sets the name of the resource that contains the symbols (pdb).
+        /// </summary>
+        public string SymbolsResourceName { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="IList{T}"/> of <see cref="RazorFileInfo"/>s.
+        /// </summary>
+        public IList<RazorFileInfo> FileInfos { get; } = new List<RazorFileInfo>();
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorFileInfoCollection.cs]: keep the list readonly

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorFileInfoCollectionGenerator.cs [hunk 7]
```diff
@@ -97,17 +99,17 @@ protected virtual string FileFormat
...
+            {{{{
+                {nameof(RazorFileInfo.LastModified)} = DateTime.FromFileTimeUtc({{0:D}}).ToLocalTime(),
+                {nameof(RazorFileInfo.Length)} = {{1:D}},
+                {nameof(RazorFileInfo.RelativePath)} = @""{{2}}"",
+                {nameof(RazorFileInfo.FullTypeName)} = @""{{3}}"",
+                {nameof(RazorFileInfo.Hash)} = ""{{4}}"",
+                {nameof(RazorFileInfo.HashAlgorithmVersion)} = {{5}},
+            }}}};
+            {nameof(RazorFileInfoCollection.FileInfos)}.Add(info);
 ";
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorFileInfoCollectionGenerator.cs]: nice, thank you C# 6

**Comment by yishaigalatzer** [comment]
:watch: looks like we need to follow up in person. Overall looks good though

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/PrecompiledViewInfo.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/PrecompiledViewInfo.cs]: Since we now attempt to read resources as assemblies given a `RazorFileInfoCollection`, our previous approach of passing the `IEnumerable<RazorFIleInfoCollection>` to `CompilerCache` ctor was being problematic. The options were having a virtual method in `CompilerCache` that we could override in the test `GetCacheEntriesFromCollections()` or refactoring the view discovery out of the `CompilerCache`. This seemed like a cleaner choice.

**Comment by pranavkm** [comment]
@yishaigalatzer updated
